### PR TITLE
Add checker conformance tests for type parameter deduction for type(t)

### PR DESCRIPTION
**Pull request by l46kok**

### CODE REVIEW

File: tests/simple/testdata/type_deduction.textproto [hunk 1]
```diff
@@ -951,3 +951,169 @@ section {
...
+        type {
+          primitive: INT64
+        }
+      }
+    }
+  }
+  test {
+    name: "type_map_erasure_comparison"
+    check_only: true
+    expr: "type({}) == map"
```
Comment (jnthntatum) [inline tests/simple/testdata/type_deduction.textproto:1078]: could you add explicit tests for the edge cases around free parameters?

I don't think these are super practical expressions but maybe better to document the weirdness than not?

It's also a bit of a can of worms so it might be fine to leave unspecified for now until we sort out the other issues with type unification.

I think the updates in C++ and java have us with something like:

```
[type([]), int]  -> list(dyn)

[] + [[type(1)]] + [[type([])]]  -> list<list<dyn>>

false ? int : type([]) -> no overload
```

Comment (l46kok) [reply tests/simple/testdata/type_deduction.textproto:1078]: Let me send these out as a separate PR, the `[] + [[type(1)]] + [[type([])]]` case especially requires further discussion (all of the stacks are divergent, so we need to decide on a behavior)

Comment (TristonianJones) [reply tests/simple/testdata/type_deduction.textproto:1078]: I seem to recall empty iteration ranges within nested comprehensions have an odd effect on type unification as well

**Review by TristonianJones — DISMISSED** [review]

File: tests/simple/testdata/type_deduction.textproto (hunk 1 above)
Comment (l46kok) [reply tests/simple/testdata/type_deduction.textproto:1078]: I've added some test cases around empty iteration ranges, PTAL

File: tests/simple/testdata/type_deduction.textproto [hunk 2]
```diff
@@ -951,3 +951,207 @@ section {
...
+    name: "empty_list_range_nested_list_comprehension"
+    check_only: true
+    expr: "[].map(x, x.map(y, y))"
+    typed_result {
+      deduced_type {
+        list_type { elem_type { list_type { elem_type { dyn {} } } } }
+      }
+    }
+  }
+  test {
```
Comment (TristonianJones) [inline tests/simple/testdata/type_deduction.textproto:1136]: Do we want to test `[].map(i, i.map(j, j > 1)) == [[]]`  and see what happens?

Comment (l46kok) [reply tests/simple/testdata/type_deduction.textproto:1136]: SGTM, added

**Review by jnthntatum — DISMISSED** [review]

**Review by jnthntatum — APPROVED** [review]
